### ISSUE TITLE
Add new "query" or "condition" parameter to stream configurations.

### ISSUE THREAD

**Issue by slotix** [slotix/dbconvert-streams-public#6]
By default, the source reader ingests all records from each table specified in the filter or from all tables if not. So, in this case, the following SQL query:

```sql
SELECT * FROM oc_filecache;
``` 
is passed to the routine that reads data from the source database. We would need to add a new parameter to enable filtering of tables, like:

```sql
SELECT * FROM oc_filecache
ORDER BY file_id
LIMIT 3042 OFFSET 2055; 
``` 

or 

```sql
SELECT * FROM oc_filecache
WHERE storage >10 
LIMIT 3042
```

**Comment by slotix** [comment]
To achieve the desired functionality, it is proposed to add a `query` parameter to each table definition in the filter. This `query` parameter allows specifying a custom SQL query for each table. Here's an example of a configuration file:

```json
{
  "source": {
    "type": "mysql",
    "mode": "convert",
    "connection": "root:pa2sW0rd@tcp(0.0.0.0:3306)/file",
    "dataBundleSize": 40,
    "reportingInterval": 5,
    "filter": {
      "tables": [
        {
          "name": "oc_filecache",
          "query": "SELECT * FROM oc_filecache ORDER BY file_id LIMIT 3042 OFFSET 2055"
        },
        {
          "name": "another_table",
          "query": "SELECT * FROM another_table WHERE storage > 10 LIMIT 3042"
        },
        {
          "name": "one_more_table"
        }
      ]
    }
  },
  "target": {
    "type": "postgresql",
    "connection": "postgres://postgres:veryStrongPassword@0.0.0.0:5432/postgres",
    "reportingInterval": 5
  }
}
```

In this example:

- Each table definition in the filter section includes a `query` parameter.
- If not, an ordinary `SELECT * FROM table;` will be used.
 
- The "query" parameter contains the custom SQL query for each specific table.
You can customize each query according to your requirements, such as adding conditions, ordering, limiting, etc.

This way, the source reader will use the specified custom queries for each table when retrieving data from the source MySQL database. Adjust the queries as needed for your specific use cases.

**Comment by slotix** [comment]
## Variants : 


### Variant 1: Using a separate configuration file for queries

```json
{
  "source": {
    "type": "mysql",
    "mode": "convert",
    "connection": "root:pa2sW0rd@tcp(0.0.0.0:3306)/file",
    "dataBundleSize": 100,
    "reportingInterval": 5,
    "filter": {
      "tables": [
        {
          "name": "oc_filecache",
          "query_file": "oc_filecache_query.sql"
        },
        {
          "name": "another_table",
          "query_file": "another_table_query.sql"
        }
      ]
    }
  },
  "target": {
    "type": "postgresql",
    "connection": "postgres://postgres:veryStrongPassword@0.0.0.0:5432/postgres",
    "reportingInterval": 5
  }
}
```

In this variant, you can store your custom queries in separate SQL files (oc_filecache_query.sql, another_table_query.sql, etc.).


### Variant 2: Using inline parameters.

```json
{
  "source": {
    "type": "mysql",
    "mode": "convert",
    "connection": "root:pa2sW0rd@tcp(0.0.0.0:3306)/file",
    "dataBundleSize": 100,
    "reportingInterval": 5,
    "filter": {
      "tables": [
        {
          "name": "oc_filecache",
          "query": "SELECT * FROM oc_filecache WHERE storage > :storage_limit LIMIT :limit",
          "parameters": {
            "storage_limit": 10,
            "limit": 3042
          }
        },
      ]
    }
  },
  "target": {
    "type": "postgresql",
    "connection": "postgres://postgres:veryStrongPassword@0.0.0.0:5432/postgres",
    "reportingInterval": 5
  }
}
```

In this variant, you can use inline parameters in your queries and define the parameter values within the configuration.

### Variant 3: Using a generic condition field

```json
{
  "source": {
    "type": "mysql",
    "mode": "convert",
    "connection": "root:pa2sW0rd@tcp(0.0.0.0:3306)/file",
    "dataBundleSize": 100,
    "reportingInterval": 5,
    "filter": {
      "tables": [
        {
          "name": "oc_filecache",
          "condition": "WHERE storage > 10 ORDER BY file_id LIMIT 3042 OFFSET 2055"
        },
        {
          "name": "another_table",
          "condition": "WHERE timestamp > '2023-01-01' ORDER BY timestamp LIMIT 100"
        }
      ]
    }
  },
  "target": {
   "type": "postgresql",
    "connection": "postgres://postgres:veryStrongPassword@0.0.0.0:5432/postgres",
    "reportingInterval": 5
  }
}
```
In this variant, you use a generic "condition" field to specify additional conditions for each table.

**Comment by slotix** [comment]
## What are the pros and cons between `query` and `condition`?

### Using "query":

#### Pros:

Explicit and Specific:
The "query" field allows to be explicit and specific about the entire SQL statement.
You have complete control over the `SELECT` statement, including the columns selected, joins, and any other clauses. 
Additionally, it is possible to feed very large tables part by part.

#### Cons:

Repetition:
If you have similar conditions across multiple tables, there might be some repetition in the queries.
Changes to the common conditions would require updates to each query individually.


###Using "condition":

####Pros:

Modularity:
The "condition" field allows for a more modular approach. You can separate common conditions from the `SELECT` statement.
Changes to common conditions can be applied uniformly across multiple tables.

####Cons:

Less Explicit:
The "condition" field might be less explicit than providing a complete `SELECT` statement.
Someone reading the configuration might need to refer to the actual query logic elsewhere.

**Comment by slotix** [comment]
Custom query support for source tables has been implemented in the latest 0.8.3 version of DBS. Explore the release details at [DBS Releases](https://stream.dbconvert.com/dbs-releases#version-0-8-3).
